File: host/potluck/potluck/value.py
"""Typed values and readings -- ARCHITECTURE.md section 4 rule 2, host side.

Mirrors firmware/components/pot_ns/include/pot/value.hpp. Second independent implementation, for
the same reason as the frame codec and the serial framing: a wire type that exists once can only be
tested against itself.

Section 4 rule 2 is carried across the language boundary, not just the wire. `Reading` here has no
attribute that hands back the bare number:

    >>> r.value            # the Value, which knows its own type
    >>> r.number()         # (number, quality) -- never one without the other
    >>> r.usable()         # False for UNAVAILABLE and NO_DATA

A `float(reading)` would be the exact bug the section exists to prevent -- a location-transparent
read whose age got dropped on the floor -- so it is not implemented, and `__float__` raising is
deliberate rather than an oversight.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from enum import IntEnum

VALUE_BYTES_MAX = 8


class ValueType(IntEnum):
    NONE = 0
    BOOL = 1
    I32 = 2
    U32 = 3
    F32 = 4
    I64 = 5
    U64 = 6
    F64 = 7
    BYTES = 8

    @classmethod
    def name_of(cls, v: int) -> str:
        try:
            return cls(v).name
        except ValueError:
            return f"type-{v}"


#: struct format and width per type. BYTES and NONE have no fixed width.
_FMT = {
    ValueType.BOOL: ("<B", 1),
    ValueType.I32: ("<i", 4),
    ValueType.U32: ("<I", 4),
    ValueType.F32: ("<f", 4),
    ValueType.I64: ("<q", 8),
    ValueType.U64: ("<Q", 8),
    ValueType.F64: ("<d", 8),
}


def value_type_size(t: int) -> int:
    """Fixed width in bytes, or 0 for NONE and BYTES (which are length-delimited)."""
    entry = _FMT.get(ValueType(t) if t in set(ValueType) else None)  # type: ignore[arg-type]
    return entry[1] if entry else 0


class Unit(IntEnum):
    NONE = 0
    RATIO = 1
    PERCENT = 2
    METRE = 10
    MILLIMETRE = 11
    RADIAN = 20
    DEGREE = 21
    METRE_PER_SECOND2 = 30
    RADIAN_PER_SECOND = 31
    NEWTON = 40
    NEWTON_METRE = 41
    CELSIUS = 50
    KELVIN = 51
    PASCAL = 60
    VOLT = 70
    MILLIVOLT = 71
    AMPERE = 72
    MILLIAMPERE = 73
    WATT = 74
    SECOND = 80
    MILLISECOND = 81
    MICROSECOND = 82
    HERTZ = 90
    COUNT = 100
    BYTE = 101
    RAW = 200

    @classmethod
    def name_of(cls, v: int) -> str:
        try:
            return cls(v).name
        except ValueError:
            return f"unit-{v}"


#: How a unit is written next to a number. Empty for the dimensionless ones, because "42 NONE" is
#: worse than "42".
_UNIT_SUFFIX = {
    Unit.NONE: "",
    Unit.RATIO: "",
    Unit.COUNT: "",
    Unit.PERCENT: "%",
    Unit.METRE: "m",
    Unit.MILLIMETRE: "mm",
    Unit.RADIAN: "rad",
    Unit.DEGREE: "deg",
    Unit.METRE_PER_SECOND2: "m/s2",
    Unit.RADIAN_PER_SECOND: "rad/s",
    Unit.NEWTON: "N",
    Unit.NEWTON_METRE: "Nm",
    Unit.CELSIUS: "C",
    Unit.KELVIN: "K",
    Unit.PASCAL: "Pa",
    Unit.VOLT: "V",
    Unit.MILLIVOLT: "mV",
    Unit.AMPERE: "A",
    Unit.MILLIAMPERE: "mA",
    Unit.WATT: "W",
    Unit.SECOND: "s",
    Unit.MILLISECOND: "ms",
    Unit.MICROSECOND: "us",
    Unit.HERTZ: "Hz",
    Unit.BYTE: "B",
    Unit.RAW: "raw",
}


def unit_suffix(u: int) -> str:
    try:
        return _UNIT_SUFFIX.get(Unit(u), Unit.name_of(u))
    except ValueError:
        return f"unit-{u}"


class Quality(IntEnum):
    GOOD = 0
    STALE = 1
    UNAVAILABLE = 2
    NO_DATA = 3
    FAULTY = 4

    @classmethod
    def name_of(cls, v: int) -> str:
        try:
            return cls(v).name
        except ValueError:
            return f"quality-{v}"


def quality_has_value(q: int) -> bool:
    """STALE is included, deliberately -- that is section 4 rule 2's whole point.

    Anything that must refuse old data says so explicitly instead of leaning on this.
    """
    return q in (Quality.GOOD, Quality.STALE)


class StalenessPolicy(IntEnum):
    INFORMATIVE = 0
    STRICT = 1


class ResourceKind(IntEnum):
    SAMPLED = 0
    EVENT = 1


class Access(IntEnum):
    READ = 1
    WRITE = 2
    READ_WRITE = 3


class NsError(IntEnum):
    """Mirrors pot::NsError. The names are what potctl prints, so they are the node's words."""

    OK = 0
    NOT_FOUND = 1
    FULL = 2
    HASH_COLLISION = 3
    TYPE_MISMATCH = 4
    NOT_WRITABLE = 5
    NOT_READABLE = 6
    WRONG_OWNER = 7
    EVENT_NOT_CACHED = 8

    @classmethod
    def name_of(cls, v: int) -> str:
        try:
            return cls(v).name
        except ValueError:
            return f"ns-error-{v}"


@dataclass(frozen=True)
class Value:
    """A typed value exactly as the wire carries it: type, length, and 8 raw little-endian bytes."""

    type: int = int(ValueType.NONE)
    raw: bytes = b""

    # --- constructors ---------------------------------------------------------------------------
    @staticmethod
    def of_bool(v: bool) -> Value:
        return Value(int(ValueType.BOOL), bytes((1 if v else 0,)))

    @staticmethod
    def of_i32(v: int) -> Value:
        return Value(int(ValueType.I32), struct.pack("<i", v))

    @staticmethod
    def of_u32(v: int) -> Value:
        return Value(int(ValueType.U32), struct.pack("<I", v))

    @staticmethod
    def of_f32(v: float) -> Value:
        return Value(int(ValueType.F32), struct.pack("<f", v))

    @staticmethod
    def of_i64(v: int) -> Value:
        return Value(int(ValueType.I64), struct.pack("<q", v))

    @staticmethod
    def of_u64(v: int) -> Value:
        return Value(int(ValueType.U64), struct.pack("<Q", v))

    @staticmethod
    def of_f64(v: float) -> Value:
        return Value(int(ValueType.F64), struct.pack("<d", v))

    @staticmethod
    def of_bytes(b: bytes) -> Value:
        if len(b) > VALUE_BYTES_MAX:
            raise ValueError(f"a Value holds at most {VALUE_BYTES_MAX} bytes, got {len(b)}")
        return Value(int(ValueType.BYTES), bytes(b))

    # --- accessors ------------------------------------------------------------------------------
    @property
    def len(self) -> int:
        return len(self.raw)

    def number(self) -> int | float | bool | None:
        """The number, or None for NONE and BYTES. Type-faithful: no reinterpretation.

        A short `raw` returns None rather than zero-padding. A truncated value is a wire fault and
        inventing the missing bytes would turn it into a plausible reading.
        """
        t = self.type
        entry = _FMT.get(ValueType(t)) if t in set(ValueType) else None  # type: ignore[arg-type]
        if entry is None:
            return None
        fmt, width = entry
        if len(self.raw) < width:
            return None
        v = struct.unpack(fmt, self.raw[:width])[0]
        return bool(v) if t == ValueType.BOOL else v

    def to_wire(self) -> tuple[int, int, bytes]:
        """(type, len, 8 bytes zero-padded) -- what WritePayload and ReplyPayload carry."""
        return self.type, len(self.raw), self.raw.ljust(VALUE_BYTES_MAX, b"\x00")

    @staticmethod
    def from_wire(vtype: int, vlen: int, raw: bytes) -> Value:
        if vlen > VALUE_BYTES_MAX:
            vlen = VALUE_BYTES_MAX
        return Value(vtype, bytes(raw[:vlen]))

    def format(self) -> str:
        if self.type == ValueType.NONE:
            return "none"
        if self.type == ValueType.BYTES:
            return "0x" + self.raw.hex() if self.raw else "0x"
        n = self.number()
        if n is None:
            return f"<{ValueType.name_of(self.type)} truncated {len(self.raw)}B>"
        if isinstance(n, bool):
            return "true" if n else "false"
        if isinstance(n, float):
            return f"{n:g}"
        return str(n)

    def json(self):
        """A JSON-safe rendering: the number itself where there is one, else a hex string."""
        if self.type == ValueType.NONE:
            return None
        if self.type == ValueType.BYTES:
            return "0x" + self.raw.hex()
        return self.number()

    def __str__(self) -> str:
        return self.format()


@dataclass
class Reading:
    """Section 4 rule 2's tuple, entire. No attribute yields the number on its own."""

    value: Value = field(default_factory=Value)
    unit: int = int(Unit.NONE)
    timestamp_ms: int = 0
    age_ms: int = 0
    latency_class: int = 4
    quality: int = int(Quality.NO_DATA)

    def usable(self) -> bool:
        return quality_has_value(self.quality)

    def number(self) -> tuple[int | float | bool | None, int]:
        """(number, quality). The only way to the number, and the quality comes with it."""
        if not self.usable():
            return None, self.quality
        return self.value.number(), self.quality

    def __float__(self):
        raise TypeError(
            "a Reading does not convert to a float: section 4 rule 2 forbids handing back a bare "
            "value. Use reading.number(), which returns (number, quality)."
        )

    def format(self) -> str:
        """One line carrying every element of the tuple. Matches Reading::format() in the firmware."""
        q = Quality.name_of(self.quality)
        if not self.usable():
            return f"{q} (age {self.age_ms} ms, class L{self.latency_class})"
        suffix = unit_suffix(self.unit)
        num = self.value.format()
        shown = f"{num} {suffix}".rstrip() if suffix else num
        return f"{shown}  [{q}, age {self.age_ms} ms, ts {self.timestamp_ms}, class L{self.latency_class}]"

    def __str__(self) -> str:
        return self.format()

File: firmware/components/pot_ns/src/value.cpp
// Typed values — see value.hpp and ARCHITECTURE.md §4 rule 2.

#include "pot/value.hpp"

#include <cstdio>
#include <cstring>

namespace pot {

const char* value_type_str(ValueType t) {
    switch (t) {
        case ValueType::None: return "none";
        case ValueType::Bool: return "bool";
        case ValueType::I32: return "i32";
        case ValueType::U32: return "u32";
        case ValueType::F32: return "f32";
        case ValueType::I64: return "i64";
        case ValueType::U64: return "u64";
        case ValueType::F64: return "f64";
        case ValueType::Bytes: return "bytes";
    }
    return "?";
}

size_t value_type_size(ValueType t) {
    switch (t) {
        case ValueType::Bool: return 1;
        case ValueType::I32:
        case ValueType::U32:
        case ValueType::F32: return 4;
        case ValueType::I64:
        case ValueType::U64:
        case ValueType::F64: return 8;
        case ValueType::None:
        case ValueType::Bytes: return 0;
    }
    return 0;
}

const char* unit_str(Unit u) {
    switch (u) {
        case Unit::None: return "";
        case Unit::Ratio: return "";
        case Unit::Percent: return "%";
        case Unit::Metre: return "m";
        case Unit::Millimetre: return "mm";
        case Unit::Radian: return "rad";
        case Unit::Degree: return "deg";
        case Unit::MetrePerSecond2: return "m/s2";
        case Unit::RadianPerSecond: return "rad/s";
        case Unit::Newton: return "N";
        case Unit::NewtonMetre: return "Nm";
        case Unit::Celsius: return "C";
        case Unit::Kelvin: return "K";
        case Unit::Pascal: return "Pa";
        case Unit::Volt: return "V";
        case Unit::Millivolt: return "mV";
        case Unit::Ampere: return "A";
        case Unit::Milliampere: return "mA";
        case Unit::Watt: return "W";
        case Unit::Second: return "s";
        case Unit::Millisecond: return "ms";
        case Unit::Microsecond: return "us";
        case Unit::Hertz: return "Hz";
        case Unit::Count: return "count";
        case Unit::Byte: return "B";
        case Unit::Raw: return "raw";
    }
    return "?";
}

const char* quality_str(Quality q) {
    switch (q) {
        case Quality::Good: return "GOOD";
        case Quality::Stale: return "STALE";
        case Quality::Unavailable: return "UNAVAILABLE";
        case Quality::NoData: return "NO_DATA";
        case Quality::Faulty: return "FAULTY";
    }
    return "?";
}

const char* staleness_policy_str(StalenessPolicy p) {
    switch (p) {
        case StalenessPolicy::Informative: return "informative";
        case StalenessPolicy::Strict: return "strict";
    }
    return "?";
}

const char* resource_kind_str(ResourceKind k) {
    switch (k) {
        case ResourceKind::Sampled: return "sampled";
        case ResourceKind::Event: return "event";
    }
    return "?";
}

// -------------------------------------------------------------------------------------------
// Value
// -------------------------------------------------------------------------------------------

namespace {
template <typename T>
Value make(ValueType t, T v) {
    Value out;
    out.type = t;
    out.len = static_cast<uint8_t>(sizeof(T));
    std::memcpy(out.raw, &v, sizeof(T));
    return out;
}

template <typename T>
bool take(const Value& val, ValueType want, T& out) {
    if (val.type != want || val.len != sizeof(T)) {
        return false;
    }
    std::memcpy(&out, val.raw, sizeof(T));
    return true;
}
}  // namespace

Value Value::of_bool(bool v) { return make(ValueType::Bool, static_cast<uint8_t>(v ? 1 : 0)); }
Value Value::of_i32(int32_t v) { return make(ValueType::I32, v); }
Value Value::of_u32(uint32_t v) { return make(ValueType::U32, v); }
Value Value::of_f32(float v) { return make(ValueType::F32, v); }
Value Value::of_i64(int64_t v) { return make(ValueType::I64, v); }
Value Value::of_u64(uint64_t v) { return make(ValueType::U64, v); }
Value Value::of_f64(double v) { return make(ValueType::F64, v); }

Value Value::of_bytes(const uint8_t* p, uint8_t n) {
    Value out;
    out.type = ValueType::Bytes;
    out.len = (n > kValueBytesMax) ? static_cast<uint8_t>(kValueBytesMax) : n;
    if (p != nullptr && out.len > 0) {
        std::memcpy(out.raw, p, out.len);
    }
    return out;
}

bool Value::as_bool(bool& out) const {
    uint8_t b = 0;
    if (!take(*this, ValueType::Bool, b)) {
        return false;
    }
    out = b != 0;
    return true;
}
bool Value::as_i32(int32_t& out) const { return take(*this, ValueType::I32, out); }
bool Value::as_u32(uint32_t& out) const { return take(*this, ValueType::U32, out); }
bool Value::as_f32(float& out) const { return take(*this, ValueType::F32, out); }
bool Value::as_i64(int64_t& out) const { return take(*this, ValueType::I64, out); }
bool Value::as_u64(uint64_t& out) const { return take(*this, ValueType::U64, out); }
bool Value::as_f64(double& out) const { return take(*this, ValueType::F64, out); }

size_t Value::format(char* out, size_t cap) const {
    if (out == nullptr || cap == 0) {
        return 0;
    }
    int n = 0;
    switch (type) {
        case ValueType::None: n = std::snprintf(out, cap, "null"); break;
        case ValueType::Bool: {
            bool b = false;
            as_bool(b);
            n = std::snprintf(out, cap, "%s", b ? "true" : "false");
            break;
        }
        case ValueType::I32: {
            int32_t v = 0;
            as_i32(v);
            n = std::snprintf(out, cap, "%ld", static_cast<long>(v));
            break;
        }
        case ValueType::U32: {
            uint32_t v = 0;
            as_u32(v);
            n = std::snprintf(out, cap, "%lu", static_cast<unsigned long>(v));
            break;
        }
        case ValueType::F32: {
            float v = 0;
            as_f32(v);
            n = std::snprintf(out, cap, "%.6g", static_cast<double>(v));
            break;
        }
        case ValueType::I64: {
            int64_t v = 0;
            as_i64(v);
            n = std::snprintf(out, cap, "%lld", static_cast<long long>(v));
            break;
        }
        case ValueType::U64: {
            uint64_t v = 0;
            as_u64(v);
            n = std::snprintf(out, cap, "%llu", static_cast<unsigned long long>(v));
            break;
        }
        case ValueType::F64: {
            double v = 0;
            as_f64(v);
            n = std::snprintf(out, cap, "%.10g", v);
            break;
        }
        case ValueType::Bytes: {
            static const char* hex = "0123456789abcdef";
            size_t w = 0;
            for (uint8_t i = 0; i < len && w + 3 < cap; ++i) {
                out[w++] = hex[raw[i] >> 4];
                out[w++] = hex[raw[i] & 0x0F];
            }
            out[w] = '\0';
            return w;
        }
    }
    if (n < 0) {
        out[0] = '\0';
        return 0;
    }
    return (static_cast<size_t>(n) < cap) ? static_cast<size_t>(n) : (cap - 1);
}

// -------------------------------------------------------------------------------------------
// §4 rule 2, in one function
// -------------------------------------------------------------------------------------------

Quality classify(uint32_t age_ms, uint32_t staleness_bound_ms, StalenessPolicy policy,
                 bool has_data, bool owner_alive) {
    // Order matters. An unreachable owner outranks everything: whatever is cached, we cannot say it
    // still reflects reality, and §8.2 has already declared the node dead.
    if (!owner_alive) {
        return Quality::Unavailable;
    }
    if (!has_data) {
        // Nothing has ever been written. Distinct from Unavailable — the resource is fine, there is
        // simply nothing to be stale about, and a caller waiting for first data wants to know that.
        return Quality::NoData;
    }
    if (staleness_bound_ms == 0 || age_ms <= staleness_bound_ms) {
        return Quality::Good;
    }
    // Past the bound. §4 rule 2: informative delivers it marked, strict delivers nothing. The one
    // thing neither does is hand it back as Good.
    return (policy == StalenessPolicy::Strict) ? Quality::Unavailable : Quality::Stale;
}

size_t Reading::format(char* out, size_t cap) const {
    if (out == nullptr || cap == 0) {
        return 0;
    }
    char vbuf[40];
    value.format(vbuf, sizeof(vbuf));
    // Every element of §4's tuple, every time: value, unit, timestamp, age, class, quality. Tests
    // assert that none of them can be omitted, because a formatter that drops the age is exactly
    // the silent stale read §4 exists to prevent.
    const int n = std::snprintf(out, cap, "%s%s%s ts=%lu age=%lums L%u %s",
                                quality_has_value(quality) ? vbuf : "-",
                                (quality_has_value(quality) && unit_str(unit)[0]) ? " " : "",
                                quality_has_value(quality) ? unit_str(unit) : "",
                                static_cast<unsigned long>(timestamp_ms),
                                static_cast<unsigned long>(age_ms),
                                static_cast<unsigned>(latency_class), quality_str(quality));
    if (n < 0) {
        out[0] = '\0';
        return 0;
    }
    return (static_cast<size_t>(n) < cap) ? static_cast<size_t>(n) : (cap - 1);
}

}  // namespace pot
